# Titanic Survival

# Homework submission for Leena Godbole (IAC-621)

### Women and children first?

#### Grading:


- Code: 90 pts
- Markdown Documentation: 10 pts


We are going to study the survival rate of passengers on titanic and what variables affected survival.

***1. Load the dataset in `titanic.xls`. It contains data on all the passengers that travelled on the Titanic. Show the head of 10 rows (10 pts)***

In [1]:
# libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# settings
pd.set_option('display.max_rows', None) 
pd.set_option('display.max_columns', 15)

df = pd.read_excel('titanic.xls')

print(df.shape)

(1309, 14)


In [2]:
df.head(10)

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
5       1         1                              Anderson, Mr. Harry    male   
6       1         1                Andrews, Miss. Kornelia Theodosia  female   
7       1         0                           Andrews, Mr. Thomas Jr    male   
8       1         1    Appleton, Mrs. Edward Dale (Charlotte Lamson)  female   
9       1         0                          Artagaveytia, Mr. Ramon    male   

       age  sibsp  parch    ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0     24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2    113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2    113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2    113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2    113781  151.5500  C22 C26        S  NaN    NaN   
5  48.0000      0      0     19952   26.5500      E12        S    3    NaN   
6  63.0000      1      0     13502   77.9583       D7        S   10    NaN   
7  39.0000      0      0    112050    0.0000      A36        S  NaN    NaN   
8  53.0000      2      0     11769   51.4792     C101        S    D    NaN   
9  71.0000      0      0  PC 17609   49.5042      NaN        C  NaN   22.0   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON  
5                     New York, NY  
6                       Hudson, NY  
7                      Belfast, NI  
8              Bayside, Queens, NY  
9              Montevideo, Uruguay

In [3]:
df.tail(10)

pclass  survived                                     name     sex   age  \
1299       3         0                      Yasbeck, Mr. Antoni    male  27.0   
1300       3         1  Yasbeck, Mrs. Antoni (Selini Alexander)  female  15.0   
1301       3         0                     Youseff, Mr. Gerious    male  45.5   
1302       3         0                        Yousif, Mr. Wazli    male   NaN   
1303       3         0                    Yousseff, Mr. Gerious    male   NaN   
1304       3         0                     Zabour, Miss. Hileni  female  14.5   
1305       3         0                    Zabour, Miss. Thamine  female   NaN   
1306       3         0                Zakarian, Mr. Mapriededer    male  26.5   
1307       3         0                      Zakarian, Mr. Ortin    male  27.0   
1308       3         0                       Zimmerman, Mr. Leo    male  29.0   

      sibsp  parch  ticket     fare cabin embarked boat   body home.dest  
1299      1      0    2659  14.4542   NaN        C    C    NaN       NaN  
1300      1      0    2659  14.4542   NaN        C  NaN    NaN       NaN  
1301      0      0    2628   7.2250   NaN        C  NaN  312.0       NaN  
1302      0      0    2647   7.2250   NaN        C  NaN    NaN       NaN  
1303      0      0    2627  14.4583   NaN        C  NaN    NaN       NaN  
1304      1      0    2665  14.4542   NaN        C  NaN  328.0       NaN  
1305      1      0    2665  14.4542   NaN        C  NaN    NaN       NaN  
1306      0      0    2656   7.2250   NaN        C  NaN  304.0       NaN  
1307      0      0    2670   7.2250   NaN        C  NaN    NaN       NaN  
1308      0      0  315082   7.8750   NaN        S  NaN    NaN       NaN

*** 2. Use the `groupby` method to calculate the proportion of passengers that survived by sex. (20 pts)*** 

In [4]:
femaleTotal = df['sex'].value_counts()['female'] # counting total amount of females in df
maleTotal = df['sex'].value_counts()['male'] # counting total amount of males in df

femaleTotal+maleTotal # ensuring that the total matches the shape of the entire dataframe


df2 = df.groupby('sex').sum() # grouping by sex and summing up values to get total survived

df2['total'] = [femaleTotal, maleTotal]

     
df2['SurvivalRate'] = df2['survived'] / df2['total']

toDrop = ['sibsp', 'parch', 'fare', 'body', 'age']

df2.drop(toDrop, inplace = True, axis = 1)

print(df2.to_markdown())



| sex    |   pclass |   survived |   total |   SurvivalRate |
|:-------|---------:|-----------:|--------:|---------------:|
| female |     1004 |        339 |     466 |       0.727468 |
| male   |     2000 |        161 |     843 |       0.190985 |


*** 3. Calculate the same proportion, but by pclass and sex. (30 pts)*** 

In [5]:
# Calculations for pclass and sex

df['count'] = 1 # adding a count variable to the original dataframe with 1 in every row.  
                #That way when I do my groupbys, I will get the total individuals in each group
    
df3 = df.groupby(['pclass', 'sex']).sum().reset_index() # grouping by sex and summing up values to get total survived

    
df3['SurvivalRate'] = df3['survived'] / df3['count']


df3.drop(toDrop, inplace = True, axis = 1)
df3 = df3.sort_values(by = 'SurvivalRate', ascending=False)


print(df3.to_markdown())



|    |   pclass | sex    |   survived |   count |   SurvivalRate |
|---:|---------:|:-------|-----------:|--------:|---------------:|
|  0 |        1 | female |        139 |     144 |       0.965278 |
|  2 |        2 | female |         94 |     106 |       0.886792 |
|  4 |        3 | female |        106 |     216 |       0.490741 |
|  1 |        1 | male   |         61 |     179 |       0.340782 |
|  5 |        3 | male   |         75 |     493 |       0.15213  |
|  3 |        2 | male   |         25 |     171 |       0.146199 |


*** 4. Create age categories: children (under 14 years), adolescents (14-20), adult (21-64), and senior(65+), and calculate survival proportions by age category, pclass and sex. (30 pts)***

In [6]:
# Creating the age tiers

tiers = [
    (df['age'] < 14),
    (df['age'] >= 14) & (df['age']<=20),
    (df['age']  >= 21) & (df['age']<=64),
    (df['age'] >= 65)
]
     
names = ['children (under 14)', 'adolescents (14-20)', 'adult (21-64)', 'senior (65+)']


df['Age Category'] = np.select(tiers, names)

df.head(50)


pclass  survived                                               name  \
0        1         1                      Allen, Miss. Elisabeth Walton   
1        1         1                     Allison, Master. Hudson Trevor   
2        1         0                       Allison, Miss. Helen Loraine   
3        1         0               Allison, Mr. Hudson Joshua Creighton   
4        1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5        1         1                                Anderson, Mr. Harry   
6        1         1                  Andrews, Miss. Kornelia Theodosia   
7        1         0                             Andrews, Mr. Thomas Jr   
8        1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9        1         0                            Artagaveytia, Mr. Ramon   
10       1         0                             Astor, Col. John Jacob   
11       1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12       1         1                      Aubart, Mme. Leontine Pauline   
13       1         1                       Barber, Miss. Ellen "Nellie"   
14       1         1               Barkworth, Mr. Algernon Henry Wilson   
15       1         0                                Baumann, Mr. John D   
16       1         0                           Baxter, Mr. Quigg Edmond   
17       1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18       1         1                              Bazzani, Miss. Albina   
19       1         0                               Beattie, Mr. Thomson   
20       1         1                      Beckwith, Mr. Richard Leonard   
21       1         1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)   
22       1         1                              Behr, Mr. Karl Howell   
23       1         1                              Bidois, Miss. Rosalie   
24       1         1                                  Bird, Miss. Ellen   
25       1         0                                Birnbaum, Mr. Jakob   
26       1         1                            Bishop, Mr. Dickinson H   
27       1         1            Bishop, Mrs. Dickinson H (Helen Walton)   
28       1         1                             Bissette, Miss. Amelia   
29       1         1          Bjornstrom-Steffansson, Mr. Mauritz Hakan   
30       1         0                       Blackwell, Mr. Stephen Weart   
31       1         1                                   Blank, Mr. Henry   
32       1         1                            Bonnell, Miss. Caroline   
33       1         1                           Bonnell, Miss. Elizabeth   
34       1         0                           Borebank, Mr. John James   
35       1         1                           Bowen, Miss. Grace Scott   
36       1         1                        Bowerman, Miss. Elsie Edith   
37       1         1      Bradley, Mr. George ("George Arthur Brayton")   
38       1         0                            Brady, Mr. John Bertram   
39       1         0                                 Brandeis, Mr. Emil   
40       1         0                          Brewe, Dr. Arthur Jackson   
41       1         1          Brown, Mrs. James Joseph (Margaret Tobin)   
42       1         1     Brown, Mrs. John Murray (Caroline Lane Lamson)   
43       1         1    Bucknell, Mrs. William Robert (Emma Eliza Ward)   
44       1         1                    Burns, Miss. Elizabeth Margaret   
45       1         0                  Butt, Major. Archibald Willingham   
46       1         0                              Cairns, Mr. Alexander   
47       1         1                  Calderhead, Mr. Edward Pennington   
48       1         1   Candee, Mrs. Edward (Helen Churchill Hungerford)   
49       1         1                 Cardeza, Mr. Thomas Drake Martinez   

       sex      age  sibsp  parch  ...        cabin  embarked boat   body  \
0   female  29.0000      0      0  ...           B5         S    2    NaN   
1     male   0.9167     

In [10]:
# Calculations for pclass, age category, and sex

df['count'] = 1 # adding a count variable to the original dataframe with 1 in every row.  
                #That way when I do my groupbys, I will get the total individuals in each group
    
df4 = df.groupby(['pclass', 'sex', 'Age Category']).sum().reset_index() # grouping by sex and summing up values to get total survived

    
df4['SurvivalRate'] = df4['survived'] / df4['count']
df4['Age Category'].replace(to_replace='0', value="unknown", inplace = True)

df4.drop(toDrop, inplace = True, axis = 1)

#print(df4.to_markdown(tablefmt="grid"))

df4 = df4.sort_values(by = 'SurvivalRate', ascending=False)

df4['pclass'] = df4['pclass'].astype(str)


#df4['Labels'] = 'Class: ' + df4.pclass + ' , Sex: ' + df.sex + ' , Age Category: ' + df['Age Category']
df4.sort_values("SurvivalRate")

#df4.head(100)
print(df4.to_markdown())


|    |   pclass | sex    | Age Category        |   survived |   count |   SurvivalRate |
|---:|---------:|:-------|:--------------------|-----------:|--------:|---------------:|
|  0 |        1 | female | unknown             |         11 |      11 |       1        |
| 17 |        2 | male   | children (under 14) |         11 |      11 |       1        |
|  4 |        1 | female | senior (65+)        |          1 |       1 |       1        |
|  1 |        1 | female | adolescents (14-20) |         15 |      15 |       1        |
| 13 |        2 | female | children (under 14) |         14 |      14 |       1        |
|  8 |        1 | male   | children (under 14) |          5 |       5 |       1        |
|  2 |        1 | female | adult (21-64)       |        112 |     116 |       0.965517 |
| 11 |        2 | female | adolescents (14-20) |         12 |      13 |       0.923077 |
| 12 |        2 | female | adult (21-64)       |         66 |      76 |       0.868421 |
| 10 |        2 | fem

##### 10 points for markdown, explain your result, and derive some conclusion.

## Analysis of Titanic Survival Rates


Below is an analysis of survival rates on the titanic by different groups of passengers.

Table 1 shows the survival rate for males vs females on the titanic.  This illustrates the cultural norm of saving women in dangerous situations by the vast difference in the survival rate between the two sexes.  Men boarded the life rafts last, and only so much room remained for survirors.   


## Table 1


| sex    |   Class  |   survived |   total |   SurvivalRate |
|:-------|---------:|-----------:|--------:|---------------:|
| female |     1004 |        339 |     466 |       0.727468 |
| male   |     2000 |        161 |     843 |       0.190985 |

Be looking at different parameters within the data, we can gain more insight on who survived the titanic.  Income and status can be derived from what class a passenger held a ticket for.  First class could only be afforded by the very wealthy of the time.  Perhaps survival outcomes were better if you were rich?  Table 2 shows the data grouped by sex and ticket class.  it illustrates that there was indid greater survival outcomes if you were wealthy. The highest survival rate was amoung first clas females, while the lowest survival rates were amount the second and third class males.  However, you can still see that if you were female, with any ticket class, your survival rate was better than males.


## Table 2

|    |    class | sex    |   survived |   count |   SurvivalRate |
|---:|---------:|:-------|-----------:|--------:|---------------:|
|  0 |        1 | female |        139 |     144 |       0.965278 |
|  2 |        2 | female |         94 |     106 |       0.886792 |
|  4 |        3 | female |        106 |     216 |       0.490741 |
|  1 |        1 | male   |         61 |     179 |       0.340782 |
|  5 |        3 | male   |         75 |     493 |       0.15213  |
|  3 |        2 | male   |         25 |     171 |       0.146199 |

The last parameter we will look at is the affects of age group.  Women and children were given priority in line for life boats, and we see that play out Table 3, with male children in the top surviver groups.

## Table 3
|    |   pclass | sex    | Age Category        |   survived |   count |   SurvivalRate |
|---:|---------:|:-------|:--------------------|-----------:|--------:|---------------:|
|  0 |        1 | female | unknown             |         11 |      11 |       1        |
| 17 |        2 | male   | children (under 14) |         11 |      11 |       1        |
|  4 |        1 | female | senior (65+)        |          1 |       1 |       1        |
|  1 |        1 | female | adolescents (14-20) |         15 |      15 |       1        |
| 13 |        2 | female | children (under 14) |         14 |      14 |       1        |
|  8 |        1 | male   | children (under 14) |          5 |       5 |       1        |
|  2 |        1 | female | adult (21-64)       |        112 |     116 |       0.965517 |
| 11 |        2 | female | adolescents (14-20) |         12 |      13 |       0.923077 |
| 12 |        2 | female | adult (21-64)       |         66 |      76 |       0.868421 |
| 10 |        2 | female | unknown             |          2 |       3 |       0.666667 |
| 20 |        3 | female | adolescents (14-20) |         19 |      35 |       0.542857 |
| 19 |        3 | female | unknown             |         34 |      64 |       0.53125  |
| 22 |        3 | female | children (under 14) |         15 |      31 |       0.483871 |
| 21 |        3 | female | adult (21-64)       |         38 |      86 |       0.44186  |
|  7 |        1 | male   | adult (21-64)       |         46 |     134 |       0.343284 |
| 26 |        3 | male   | children (under 14) |         12 |      37 |       0.324324 |
|  5 |        1 | male   | unknown             |          8 |      28 |       0.285714 |
|  6 |        1 | male   | adolescents (14-20) |          1 |       5 |       0.2      |
| 25 |        3 | male   | adult (21-64)       |         39 |     244 |       0.159836 |
| 14 |        2 | male   | unknown             |          2 |      13 |       0.153846 |
|  9 |        1 | male   | senior (65+)        |          1 |       7 |       0.142857 |
| 24 |        3 | male   | adolescents (14-20) |          8 |      64 |       0.125    |
| 15 |        2 | male   | adolescents (14-20) |          2 |      17 |       0.117647 |
| 23 |        3 | male   | unknown             |         16 |     145 |       0.110345 |
| 16 |        2 | male   | adult (21-64)       |         10 |     128 |       0.078125 |
| 18 |        2 | male   | senior (65+)        |          0 |       2 |       0        |
|  3 |        1 | female | children (under 14) |          0 |       1 |       0        |
| 27 |        3 | male   | senior (65+)        |          0 |       3 |       0        |
